File: front-end/src/services/FbService.js
// FbService.js
// 목적 : Fb 정보를 axios 로 CRUD 를 하는 공통 함수들 정의
import http from "../http-common";

class FbService {
  // 전체 조회 (페이징)
  getAll(fb_ctitle, fb_title, page, size) {
    return http.get(
      `/fb?fb_ctitle=${fb_ctitle}&fb_title=${fb_title}&page=${page}&size=${size}`,
      {}
    );
  }
  
  // 조회수 best 5
  getFbBoardBest() {
    return http.get(`/fb/best`);
  }

  // 댓글수 best 5 
  getFbBoardRcntBest() {
    return http.get(`/fb/rcnt`);
  }

   // 최신순 5
   getFbDesc() {
    return http.get(`/fb/desc`);
  }

  // 상세페이지 조회
  getDetailAll() {
    return http.get(`/fb/all`);
  }

  // 게시번호로 조회
  get(fno) {
    return http.get(`/fb/${fno}`);
  }

  // 게시글 생성
  create(nick, fbCategory, fbTitle, fbContent, fbTag, blobFile, onUploadProgress) {
    let formData = new FormData();

    console.log(nick);
    console.log(fbCategory);
    console.log(fbTitle);
    console.log(fbContent);
    console.log(fbTag);

    console.log(blobFile);
    
    formData.append("nick", nick);
    formData.append("fbCategory", fbCategory);
    formData.append("fbTitle", fbTitle);
    formData.append("fbContent", fbContent);
    formData.append("fbTag", fbTag);
    formData.append("blobFile", blobFile);

    return http.post("/fb", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
      onUploadProgress
    });
  }
  
  // 수정
  update(
    fno,
    nick,  
    fbCategory,
    fbTitle,
    fbContent,
    fbTag,
    fbCnt,
    blobFile,
    onUploadProgress
  ) {
    let formData = new FormData();

    console.log(fno);
    console.log(nick);
    console.log(fbCategory);
    console.log(fbTitle);
    console.log(fbContent);
    console.log(fbTag);
    console.log(fbCnt);

    console.log(blobFile);

    formData.append("fno", fno);
    formData.append("nick", nick);
    formData.append("fbCategory", fbCategory);
    formData.append("fbTitle", fbTitle);
    formData.append("fbContent", fbContent);
    formData.append("fbTag", fbTag);
    formData.append("fbCnt", fbCnt);
    formData.append("blobFile", blobFile);

    return http.put(`/fb/${fno}`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
      onUploadProgress,
    });
  }

  // 삭제
  delete(fno) {
    return http.delete(`/fb/deletion/${fno}`);
  }
}

export default new FbService();

File: DPD_backend/src/main/java/com/example/simpledms/service/community/FbService.java
package com.example.simpledms.service.community;

import com.example.simpledms.dto.community.FbCategoryDto;
import com.example.simpledms.dto.community.FbCommentDto;
import com.example.simpledms.model.community.Fb;
import com.example.simpledms.repository.community.FbRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.Optional;

/**
 * packageName : com.example.simpledms.service.community
 * fileName : FbService
 * author : ds
 * date : 2022-12-19
 * description : 자유게시판 서비스 클래스
 * ===========================================================
 * DATE            AUTHOR             NOTE
 * —————————————————————————————
 * 2022-12-19         ds          최초 생성
 */
@Service
public class FbService {
    // CRUD 함수 사용을 위한 Repository 불러오기
    @Autowired
    FbRepository fbRepository;

    // 이미지 포함한 생성 메서드
    public Fb createUploadFile(String nick,
                               String fbCategory,
                               String fbTitle,
                               String fbContent,
                               String fbTag,
                               MultipartFile blobFile)
            throws IOException {

        if(blobFile != null) {
            // 업로드 파일에서 파일명 얻기
            String fileName = StringUtils.cleanPath(blobFile.getOriginalFilename());

            Fb fb = Fb.builder()
                    .nick(nick)
                    .fbCategory(fbCategory)
                    .fbTitle(fbTitle)
                    .fbContent(fbContent)
                    .fbTag(fbTag)
                    .fileName(fileName)
                    .blobFile(blobFile.getBytes())
                    .build();

            Fb createFb = fbRepository.save(fb);
            return createFb;

        } else {
            Fb fb = Fb.builder()
                    .nick(nick)
                    .fbCategory(fbCategory)
                    .fbTitle(fbTitle)
                    .fbContent(fbContent)
                    .fbTag(fbTag)
                    .fileName("")
                    .blobFile(null)
                    .build();

            Fb createFb = fbRepository.save(fb);
            return createFb;
        }
    }

    // 이미지 포함한 수정 메서드
    public Fb updateUploadFile(int fno, String nick, String fbCategory, String fbTitle, String fbContent, String fbTag, int fbCnt,
                               MultipartFile blobFile) throws IOException {

        if (blobFile != null) {
            // 업로드 파일에서 파일명 얻기
            String fileName = StringUtils.cleanPath(blobFile.getOriginalFilename());

            Fb fb = Fb.builder()
                    .fno(fno)
                    .nick(nick)
                    .fbCategory(fbCategory)
                    .fbTitle(fbTitle)
                    .fbContent(fbContent)
                    .fbTag(fbTag)
                    .fbCnt(fbCnt)
                    .blobFile(blobFile.getBytes())
                    .build();

            Fb createFb = fbRepository.save(fb);

            return createFb;
        } else {
            Fb fb = Fb.builder()
                    .fno(fno)
                    .nick(nick)
                    .fbCategory(fbCategory)
                    .fbTitle(fbTitle)
                    .fbContent(fbContent)
                    .fbTag(fbTag)
                    .fbCnt(fbCnt)
                    .blobFile(null)
                    .build();

            Fb createFb = fbRepository.save(fb);
            return createFb;
        }
    }

    // 카테고리 선택 후 검색 조회 메서드
    public Page<FbCategoryDto> selectSearchAndKeywordSearch(String fb_ctitle,
                                                            String fb_title,
                                                            Pageable pageable) {
        Page<FbCategoryDto> fbPage
                = fbRepository.selectSearchAndKeywordSearch(fb_ctitle, fb_title, pageable);

        return fbPage;
    }

    // 자유게시판 상세 게시판 메서드
    public List<Fb> findAll() {

        List<Fb> fbList = fbRepository.findAll();
        return fbList;
    }

    // 게시판 번호(fno) 로 상세조회 메서드
    public Optional<Fb> findId(int fno) {

        // findById : parameter 값 - ID, return  값 - Optional
        Optional<Fb> fbOptional = fbRepository.findById(fno);

        return fbOptional;
    }

    // 게시판 번호(fno)로 삭제하는 메서드
    public boolean removeById(int fno) {

        if (fbRepository.existsById(fno)) {
            fbRepository.deleteById(fno);
            return true;
        }
        return false;
    }

    // 상세페이지 클릭시 조회수 1상승 메서드
    // @Transactional : 해당 범위 내 메서드가 트랜잭션이 되도록 보장해준다.
    @Transactional
    public int updateCnt(Integer fno) {
        return fbRepository.updateCnt(fno);
    }

    // 조회수 Best 5 리스트 메서드
    public List<Fb> boardBestList() {
        List<Fb> list = fbRepository.boardBestList();

        return list;
    }

    // 최신순 6 리스트 메서드
    public List<Fb> boardDescList() {
        List<Fb> list = fbRepository.boardDescList();

        return list;
    }

    // 댓글수 Best 5 리스트 메서드
    public List<FbCommentDto> findBestRcnt() {
        List<FbCommentDto> list = fbRepository.findBestRcnt();

        return list;
    }
}
